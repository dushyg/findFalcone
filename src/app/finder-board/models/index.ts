export { IFalconAppState } from './falconApp.state';
export { RawPlanet, IPlanet } from './planet';
export { RawVehicle, IVehicle } from './vehicle';
export { default as PlanetChange } from './planetChange';
export { default as VehicleChange } from './vehicleChange';
export { ISearchAttempt } from './searchAttempt';
