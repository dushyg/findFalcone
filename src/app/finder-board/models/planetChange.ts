import { IPlanet } from './planet';

export default class PlanetChange {
  constructor(public widgetId: number, public newPlanetName: string) {}
}
