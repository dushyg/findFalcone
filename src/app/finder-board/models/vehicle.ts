// tslint:disable: indent
export interface IVehicle {
  name: string;
  totalNumUnits: number;
  maxDistance: number;
  speed: number;
  availNumUnits: number;
}

export interface RawVehicle {
  id?: number;
  name: string;
  total_no: number;
  max_distance: number;
  speed: number;
}
