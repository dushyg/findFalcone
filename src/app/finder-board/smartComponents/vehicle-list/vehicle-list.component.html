<div class="vehicleList">
  <div *ngFor="let vehicle of vehicleList; let i = index">
    <input
      type="radio"
      name="radio_{{ widgetId }}"
      [value]="vehicle.name"
      [ngModel]="lastSelectedVehicle.name"
      [disabled]="
        (vehicle && vehicle.maxDistance < destinationDistance) ||
        vehicle.availNumUnits == 0
      "
      id="vehicle_{{ widgetId }}_{{ i }}"
      (change)="vehicleSelected(vehicle)"
    />&nbsp;
    <label for="vehicle_{{ widgetId }}_{{ i }}">
      {{ vehicle.name + " (" + vehicle.availNumUnits + ")" }}
    </label>

    <br />
  </div>
</div>
