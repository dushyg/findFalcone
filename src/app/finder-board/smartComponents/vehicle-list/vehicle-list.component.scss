.vehicleList {
  display: inline-block;
  text-align: left;
}

input[type="radio"]:disabled {
  cursor: not-allowed;
}

input[type="radio"]:disabled + label {
  cursor: not-allowed;
  text-decoration: line-through;
  color: lightcoral;
}

input[type="radio"]:checked + label {
  font-weight: bold;
  font-style: italic;
  text-decoration: none;
  color: black;
}
