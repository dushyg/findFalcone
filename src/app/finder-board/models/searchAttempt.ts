import { IVehicle } from "./vehicle";
import { IPlanet } from "./planet";

export interface ISearchAttempt {
  searchedPlanet: string;
  vehicleUsed: string;
}
