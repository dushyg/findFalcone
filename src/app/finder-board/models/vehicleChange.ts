import { IVehicle } from './vehicle';

export default class VehicleChange {
  constructor(public widgetId: number, public newVehicleName) {}
}
