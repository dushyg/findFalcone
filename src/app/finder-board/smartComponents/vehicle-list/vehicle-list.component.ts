import {
  Component,
  OnInit,
  Input,
  OnDestroy,
  ChangeDetectionStrategy,
  ChangeDetectorRef,
} from '@angular/core';
import { IVehicle } from 'src/app/finder-board/models/vehicle';
import VehicleChange from 'src/app/finder-board/models/vehicleChange';
import { FinderFacadeService } from 'src/app/finder-board/services/finder-facade.service';
import { takeWhile, withLatestFrom } from 'rxjs/operators';

@Component({
  selector: 'app-vehicle-list',
  templateUrl: './vehicle-list.component.html',
  styleUrls: ['./vehicle-list.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class VehicleListComponent implements OnInit, OnDestroy {
  private isComponentActive = true;

  constructor(
    private finderFacadeService: FinderFacadeService,
    private changeDetector: ChangeDetectorRef
  ) {}

  public vehicleList: IVehicle[];

  @Input() public destinationDistance: number;
  @Input() public widgetId: number;

  public lastSelectedVehicle: IVehicle = {
    availNumUnits: 0,
    name: null,
  } as IVehicle;

  ngOnInit() {
    let vehicleListInitialized = false;

    this.finderFacadeService.vehicleInventory$
      .pipe(takeWhile(() => !vehicleListInitialized))
      .subscribe((initialVehicles: IVehicle[]) => {
        if (initialVehicles) {
          this.vehicleList = initialVehicles;
          vehicleListInitialized = true;
          this.changeDetector.detectChanges();
        }
      });

    this.finderFacadeService.vehicleChangedAction$
      .pipe(
        withLatestFrom(this.finderFacadeService.vehicleInventory$),
        takeWhile(() => this.isComponentActive)
      )
      .subscribe(([vehicleChange, updatedVehicles]) => {
        if (vehicleChange && vehicleChange.widgetId <= this.widgetId) {
          this.vehicleList = updatedVehicles;
          this.changeDetector.detectChanges();
        }
      });

    this.finderFacadeService.planetChangedAction$
      .pipe(
        withLatestFrom(this.finderFacadeService.vehicleInventory$),
        takeWhile(() => this.isComponentActive)
      )
      .subscribe(([planetChange, updatedVehicles]) => {
        // need to refresh the vehicle list even when the planet is chnaged for current widget
        if (planetChange && planetChange.widgetId <= this.widgetId) {
          this.vehicleList = updatedVehicles;
          if (planetChange.widgetId !== this.widgetId) {
            this.clearLastSelection();
            this.changeDetector.detectChanges();
          }
        }
      });
  }

  public vehicleSelected(vehicle: IVehicle) {
    this.lastSelectedVehicle = vehicle;

    this.finderFacadeService.vehicleChanged(
      new VehicleChange(this.widgetId, vehicle.name)
    );
  }

  clearLastSelection(): void {
    this.lastSelectedVehicle = { availNumUnits: 0, name: null } as IVehicle;
  }

  ngOnDestroy(): void {
    this.isComponentActive = false;
  }
  // tslint:disable-next-line: use-lifecycle-interface
  ngAfterViewChecked(): void {
    // console.log('VehicleListComponent ngAfterViewChecked', this.widgetId);
  }
}
