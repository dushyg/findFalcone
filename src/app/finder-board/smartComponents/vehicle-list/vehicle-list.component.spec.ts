import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { VehicleListComponent } from './vehicle-list.component';
import { IVehicle } from '../../models/vehicle';
import { createSpyObj } from '../../../testingUtility';
import { FinderFacadeService } from '../../services/finder-facade.service';
import { PlanetsService } from '../../services/planets.service';
import { VehiclesService } from '../../services/vehicles.service';
import { FalconeTokenService } from '../../services/falcone-token.service';
import { Router } from '@angular/router';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { of } from 'rxjs';

describe('VehicleListComponent', () => {
  let component: VehicleListComponent;
  let fixture: ComponentFixture<VehicleListComponent>;

  let planetServiceMock: { [key: string]: jest.Mock<any, any> };
  let vehiclesServiceMock: { [key: string]: jest.Mock<any, any> };
  let apiTokenServiceMock: { [key: string]: jest.Mock<any, any> };
  let routerServiceMock: { [key: string]: jest.Mock<any, any> };

  let vehicleListToBeReturned: IVehicle[];
  const apiTokenToBeReturned = { token: 'plmVHX' };
  beforeEach(() => {
    planetServiceMock = createSpyObj(['getAllPlanets']);
    vehiclesServiceMock = createSpyObj(['getAllVehicles']);
    apiTokenServiceMock = createSpyObj(['getFalconeFinderApiToken']);
    routerServiceMock = createSpyObj(['navigate']);

    TestBed.configureTestingModule({
      declarations: [VehicleListComponent],
      providers: [
        FinderFacadeService,
        { provide: PlanetsService, useValue: planetServiceMock },
        { provide: VehiclesService, useValue: vehiclesServiceMock },
        { provide: FalconeTokenService, useValue: apiTokenServiceMock },
        { provide: Router, useValue: routerServiceMock },
      ],
      schemas: [NO_ERRORS_SCHEMA],
    });

    vehicleListToBeReturned = [
      {
        name: 'Space pod',
        availNumUnits: 2,
        maxDistance: 200,
        speed: 2,
        totalNumUnits: 2,
      },
      {
        name: 'Space rocket',
        availNumUnits: 1,
        maxDistance: 300,
        speed: 4,
        totalNumUnits: 1,
      },
      {
        name: 'Space shuttle',
        availNumUnits: 1,
        maxDistance: 400,
        speed: 5,
        totalNumUnits: 1,
      },
      {
        name: 'Space ship',
        availNumUnits: 2,
        maxDistance: 600,
        speed: 10,
        totalNumUnits: 2,
      },
    ];
    const planetListToBeReturned = [
      { name: 'Donlon', distance: 100 },
      { name: 'Enchai', distance: 200 },
      { name: 'Jebing', distance: 300 },
      { name: 'Sapir', distance: 400 },
    ];
    planetServiceMock.getAllPlanets.mockReturnValue(of(planetListToBeReturned));
    vehiclesServiceMock.getAllVehicles.mockReturnValue(
      of(vehicleListToBeReturned)
    );
    apiTokenServiceMock.getFalconeFinderApiToken.mockReturnValue(
      of(apiTokenToBeReturned)
    );

    fixture = TestBed.createComponent(VehicleListComponent);

    component = fixture.componentInstance;
  });

  it('should create', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
  });

  it('should render radio buttons equal to number of vehicles', () => {
    fixture.componentRef.injector.get(FinderFacadeService).initializeAppData();
    fixture.detectChanges();

    const radioBtns = (fixture.nativeElement as HTMLElement).querySelectorAll(
      'input[type=radio]'
    );
    expect(radioBtns).toBeTruthy();
    expect(radioBtns.length).toBe(vehicleListToBeReturned.length);

    const nameSet = new Set<string>();
    component.vehicleList.forEach((v) => nameSet.add(v.name));
    radioBtns.forEach((radioBtn) => {
      expect(nameSet.has((radioBtn as HTMLInputElement).value)).toBeTruthy();
    });
    (expect(fixture) as any).toMatchSnapshot();
  });

  it('should render labels equal to number of vehicles', () => {
    fixture.componentRef.injector.get(FinderFacadeService).initializeAppData();
    component.widgetId = 1;
    component.destinationDistance = 100;
    fixture.detectChanges();

    const labels = (fixture.nativeElement as HTMLElement).querySelectorAll(
      'label'
    );

    expect(labels).toBeTruthy();
    expect(labels.length).toBe(vehicleListToBeReturned.length);

    let index = 0;
    labels.forEach((label) => {
      const vehicleByIndex = component.vehicleList[index++];
      expect(label.textContent).toContain(
        `${vehicleByIndex.name} (${vehicleByIndex.availNumUnits})`
      );
    });
  });
});
