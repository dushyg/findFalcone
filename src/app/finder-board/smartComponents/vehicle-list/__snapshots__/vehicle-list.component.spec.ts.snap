// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VehicleListComponent should render radio buttons equal to number of vehicles 1`] = `
<app-vehicle-list
  changeDetector={[Function ViewRef_]}
  finderFacadeService={[Function FinderFacadeService]}
  isComponentActive={[Function Boolean]}
  lastSelectedVehicle={[Function Object]}
  vehicleList={[Function Array]}
>
  <div
    class="vehicleList"
  >
    
    <div>
      <input
        id="vehicle__0"
        name="radio_"
        type="radio"
        value="Space pod"
      />
        
      <label
        for="vehicle__0"
      >
         Space pod (2) 
      </label>
      <br />
    </div>
    <div>
      <input
        id="vehicle__1"
        name="radio_"
        type="radio"
        value="Space rocket"
      />
        
      <label
        for="vehicle__1"
      >
         Space rocket (1) 
      </label>
      <br />
    </div>
    <div>
      <input
        id="vehicle__2"
        name="radio_"
        type="radio"
        value="Space shuttle"
      />
        
      <label
        for="vehicle__2"
      >
         Space shuttle (1) 
      </label>
      <br />
    </div>
    <div>
      <input
        id="vehicle__3"
        name="radio_"
        type="radio"
        value="Space ship"
      />
        
      <label
        for="vehicle__3"
      >
         Space ship (2) 
      </label>
      <br />
    </div>
  </div>
</app-vehicle-list>
`;
